File: longest-increasing-subsequence/njngwn.java
// time complexity: O(nlogn), n: nums.length (logn because of binary search)
// space complexity: O(n), n: nums.length
class Solution {
    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer> incSeqList = new ArrayList<Integer>();   // dp
        incSeqList.add(nums[0]);

        for (int num : nums) {
            if (num > incSeqList.get(incSeqList.size()-1)) {
                // add element to incSeqLit
                incSeqList.add(num);
            } else {
                int idx = Collections.binarySearch(incSeqList, num);
                if (idx < 0) {  // idx returns -(insertedPos + 1)
                    int insertedIdx = -(idx + 1);
                    incSeqList.set(insertedIdx, num);
                }
            }
        }

        return incSeqList.size();
    }
}

File: longest-increasing-subsequence/njngwn.py
class Solution:
    # Time Complexity: O(n*log(n)), n: len(nums)
    # Space Complexity: O(n), n: len(nums)
    def lengthOfLIS(self, nums: List[int]) -> int:
        seq = []

        for num in nums:
            if not seq or (seq[-1] < num):
                seq.append(num)
                continue

            # overwrite the smaller value with binary search
            left, right = 0, len(seq)
            while left < right:
                mid = (left + right) // 2
                if seq[mid] < num:
                    left = mid + 1
                else:
                    right = mid

            seq[left] = num

        return len(seq)
